### ISSUE TITLE
[RFC] Completing Quantization Support for MoE Inference in TorchAO

### ISSUE THREAD

**Issue by eryk-roch** [pytorch/ao#4939]
## Summary

Although experts in Mixture-of-Experts (MoE) models are conceptually collections of linear projections, MoE implementations frequently pack multiple expert weights into a single grouped representation (for example a 3D `(num_experts, N, K)` parameter). These weights are usually executed with grouped matrix multiplication (`grouped_mm`) kernels for efficiency. This differs from the conventional dense `nn.Linear` workflow and requires quantization to operate on grouped expert-weight representations and their execution path rather than on individual `nn.Linear.weight` tensors.

TorchAO already supports several relevant quantization formats and has grouped-MM support for some of them. The remaining gap is an end-to-end user workflow: applying a chosen format to the complete expert weights and running the model's existing MoE execution path with clear support boundaries.

This RFC builds on [RFC #4355](https://github.com/pytorch/ao/issues/4355) and serves as an umbrella for complete MoE quantization support in TorchAO across formats, backends, expert-parameter targeting, and validation. It does not prescribe a single implementation for all formats.

## Motivation

A user should be able to quantize an MoE model in the same general way as a dense model:

```python
model = AutoModelForCausalLM.from_pretrained(model_id, torch_dtype=torch.bfloat16)
quantize_(model, config=quant_config, filter_fn=select_expert_modules)
output = model(input_ids)
```

The quantized model should preserve the model's routing and execution behavior. In particular, TorchAO needs to handle:

* expert weights represented as a group rather than independent dense linear weights;
* grouped-MM execution and variable token-to-expert assignments;
* different quantization formats and their hardware constraints;
* expert parameters that are not exposed through the conventional `.weight` interface.

The user-facing requirement is simple: the requested expert weights should be quantized, the model should run through its intended path, and unsupported combinations should be reported explicitly.

## Current state

The table below covers MoE-specific support only; dense `nn.Linear` support is not listed.

| Format | Supported for MoE today | Missing for MoE |
| --- | --- | --- |
| INT4 weight-only | `Int4Tensor` and `Int4PreshuffledTensor` (CUDA, FBGEMM/MSLK kernels) accept 3D stacked expert weights in `from_hp` and implement `torch.bmm`; `Int4Tensor` also supports 3D `transpose`, `cat`, `view` (2D↔3D) and `squeeze`. `Int4TilePackedTo4dTensor` packs 3D weights and implements `aten.select.int` for expert selection. Expert parameters can be quantized via explicit `parameter_name` targeting. | No INT4 tensor subclass implements `aten._grouped_mm`, so the `grouped_mm` experts path cannot run on INT4 weights. `Int4PlainInt32Tensor` (XPU/NPU) is 2D-only: no 3D `from_hp`, expert `transpose`, `select`, `index`, `bmm` or `grouped_mm`. `Int4Tensor.slice` is 2D-only. Weights incompatible with `group_size` are left unquantized with only an info log. No automatic discovery of expert parameters that are not exposed as `.weight` (e.g. `gate_up_proj`, `down_proj`). No end-to-end test on a real MoE model. |
| FP8 row-wise | `Float8Tensor` implements `aten._grouped_mm` ([PR #4390](https://github.com/pytorch/ao/pull/4390)). With `Float8DynamicActivationFloat8WeightConfig(granularity=PerRow())`, activations are quantized dynamically and dispatched to `scaled_grouped_mm` with row-wise scales. The weight-only path falls back to dequantize + `torch._grouped_mm`. | Only `PerRow` granularity is supported; other granularities raise `NotImplementedError`. The weight-only path dequantizes the whole expert stack on every call, so it gives no runtime memory or compute benefit. Weights failing the FP8 shape check are silently left unquantized. No validated matrix of models, hardware and backends (e.g. XPU). |
| NVFP4 | `NVFP4Tensor` implements `aten._grouped_mm` ([PR #4316](https://github.com/pytorch/ao/pull/4316)) via `scaled_grouped_mm` with 1x16 block-wise and per-expert tensor-wise scales. The dynamic mode of `NVFP4DynamicActivationNVFP4WeightConfig` accepts 3D expert weights with per-expert global scales; 3D `transpose` is supported. | SM100+ only. Static calibration (`PREPARE`/`CONVERT`) and `NVFP4WeightOnlyConfig` assert 2D weights. `NVFP4Tensor` does not implement `aten.index.Tensor`, so the Transformers `batched_mm` experts path (used by `generate()` for decode) fails; the GPTQ example has to force `grouped_mm`. |
| MXFP8 / MXFP4 | `MXTensor` primitives with 2D `mm`/`linear`, `select`, `slice`, `cat`. MoE *training* support in `torchao.prototype.moe_training` (e.g. `_to_mxfp8_then_scaled_grouped_mm`), which keeps high-precision weights and quantizes dynamically per step. | `MXTensor` does not implement `aten._grouped_mm` (proposed in [PR #4419](https://github.com/pytorch/ao/pull/4419)), so `MXDynamicActivationMXWeightConfig` cannot serve pre-quantized expert weights through the `grouped_mm` path. No MXFP4 grouped-MM inference path. |

Targeting is also part of the end-to-end problem. Explicit parameter targeting exists, but model integrations need a reliable way to identify all intended expert parameters during quantization and checkpoint loading. [PR #4894](https://github.com/pytorch/ao/pull/4894) proposed this for a staged INT4 workflow and is the basis for step 2 in [Follow-up PRs](#follow-up-prs).

## Use cases

The snippets below show the intended user-facing shape, not guaranteed end-to-end workflows available today.

### INT4: Qwen3-Coder-Next and similar models

```python
model = AutoModelForCausalLM.from_pretrained(
    "Qwen/Qwen3-Coder-Next", torch_dtype=torch.float16
)
quantize_(model, config=Int4WeightOnlyConfig(), filter_fn=select_expert_modules)
```

The user wants W4A16 expert inference while leaving routing and non-expert layers unchanged. TorchAO has partial expert-parameter quantization support, but not yet a complete supported workflow that quantizes the full expert weights and runs them through the model's existing MoE path.

Ideally, INT4 would reduce expert-weight storage from roughly $2P$ BF16 bytes to $P/2$ payload bytes for $P$ parameters, before scale metadata; the current partial workflow does not provide one meaningful end-to-end quantized size.

### FP8: DeepSeek-V3 or Qwen3-MoE

```python
model = AutoModelForCausalLM.from_pretrained(
    "deepseek-ai/DeepSeek-V3", torch_dtype=torch.bfloat16
)
quantize_(
    model,
    config=Float8DynamicActivationFloat8WeightConfig(),
    filter_fn=select_expert_modules,
)
```

The user wants FP8 expert inference. FP8 grouped-MM support has been implemented in [PR #4390](https://github.com/pytorch/ao/pull/4390) for a supported row-wise workflow; remaining work is to define and validate supported model, hardware, and execution combinations rather than treating this as universal MoE support.

DeepSeek-V3 has a very large expert-weight footprint, making the memory impact of FP8 particularly significant.

### NVFP4: OLMoE and similar deployments

```python
model = AutoModelForCausalLM.from_pretrained(
    "allenai/OLMoE-1B-7B-0924", torch_dtype=torch.bfloat16
)
quantize_(
    model,
    config=NVFP4DynamicActivationNVFP4WeightConfig(),
    filter_fn=select_expert_modules,
)
```

The user wants FP4 expert inference. The repository contains a prototype NVFP4 grouped-MM path, based on [PR #4316](https://github.com/pytorch/ao/pull/4316), but it remains dependent on supported hardware and does not represent a complete, format-independent MoE workflow.

For an approximately 7B-parameter expert portion, BF16 storage is about 14 GB; an ideal FP4 payload is about 3.5 GB before scales and other overhead.

MXFP8 and MXFP4 are additional examples: the format primitives and some MoE training support exist, but a complete inference workflow for MoE grouped-MM is not yet established in the current repository.

## Relationship to RFC #4355

Since RFC #4355 was written, parts of its direction have been implemented: FP8 grouped-MM (#4390) and NVFP4 grouped-MM (#4316).

This RFC addresses the problem more broadly and acts as an umbrella for complete MoE quantization support in TorchAO. Beyond scaled grouped GEMM for activation-quantized formats, it also covers weight-only formats such as INT4, MXFP8/MXFP4 inference, expert-parameter targeting and checkpoint loading, and cross-format validation.

## Proposed direction

TorchAO should incrementally provide:

* support for quantizing the expert-weight representations used by target MoE models;
* execution of those weights through the model's existing grouped-MM or equivalent expert path;
* clear format/backend capability boundaries, including supported recipes, hardware, shapes, and fallback behavior;
* reliable integration-level parameter targeting, including checkpoint-loading workflows;
* validation of numerical accuracy, memory reduction, eager execution, compilation, and serialization for each supported capability.

The exact tensor representations, layouts, kernel implementations, dispatcher behavior, and loader APIs should be defined in the corresponding implementation PRs, not in this RFC.

## Scope and non-goals

In scope:

* quantized MoE expert weights;
* format-specific grouped-MM or equivalent execution support;
* user-visible capability and failure behavior;
* integration needed to quantize complete expert parameters.

Not in scope:

* model-specific routing or token assignment;
* designing individual kernels or backend-specific layouts;
* specifying tensor subclasses, packing, dispatchers, or checkpoint-loader internals;
* guaranteeing one format works on every device or architecture.

## Follow-up PRs

The work is split into focused PRs, in the following order:

1. **INT4 MoE enablement for `Int4PlainInt32Tensor` (XPU)** – based on [PR #4731](https://github.com/pytorch/ao/pull/4731), following the approach of #4390 and #4316: quantization of 3D stacked expert weights `(E, N, K)`, expert-stack dequantize/transpose, `select`/`index` for expert routing, `bmm`, `grouped_mm` with `offs` validation, and `torch.compile` support.
2. **Expert-parameter targeting in `quantize_`** – based on [PR #4894](https://github.com/pytorch/ao/pull/4894): MoE-aware default module matching, auto-discovery of expert parameters not exposed as `.weight`, and CPU-side checkpoint staging for loader integrations, starting with INT4.
3. **INT4 on other backends** – `grouped_mm` for `Int4Tensor`, `Int4PreshuffledTensor` and `Int4TilePackedTo4dTensor`.
4. **FP8** – granularities beyond `PerRow`, weight-only execution without full dequantization, validated model/hardware matrix.
5. **NVFP4** – 3D weights in static calibration and `NVFP4WeightOnlyConfig`, missing expert-routing ops.
6. **MXFP8/MXFP4** – `grouped_mm` inference support for `MXTensor` (see [PR #4419](https://github.com/pytorch/ao/pull/4419)).
7. **Cross-format validation** – model-level tests for eager, compile, serialization, accuracy, and memory across supported format/backend combinations.

## Research notes

This RFC is based on the current TorchAO implementation and the following references:

* [RFC #4355](https://github.com/pytorch/ao/issues/4355): broad direction for MoE inference quantization.
* [PR #4316](https://github.com/pytorch/ao/pull/4316): NVFP4 grouped-MM.
* [PR #4390](https://github.com/pytorch/ao/pull/4390): FP8 grouped-MM.
* [PR #4419](https://github.com/pytorch/ao/pull/4419): proposed MX grouped-MM inference support.
* [PR #4731](https://github.com/pytorch/ao/pull/4731): MoE-ready `Int4PlainInt32Tensor` execution on XPU (closed; basis for step 1).
* [PR #4894](https://github.com/pytorch/ao/pull/4894): MoE-aware `quantize_` with staged INT4 expert auto-targeting (closed; basis for step 2).
* Current implementations and tests for `Int4PlainInt32Tensor`, `Float8Tensor`, `NVFP4Tensor`, and `MXTensor`.

### Proposed API

```python

```

### Target hardware

None

**Comment by liangan1** [comment]
We need to reuse the https://github.com/pytorch/ao/issues/4355
